### PR TITLE
GUAC-586: Support multiple auth providers within web UI

### PR DESCRIPTION
**Pull request by mike-jumper**
This change touches practically everything in the UI and REST API, adding support for the support for multiple extensions we added in #250.

### CODE REVIEW

File: guacamole/src/main/java/org/glyptodon/guacamole/net/basic/rest/ObjectRetrievalService.java [hunk 1]
```diff
@@ -48,7 +49,7 @@
      * @param session
      *     The GuacamoleSession to retrieve the UserContext from.
      *
-     * @param identifier
+     * @param authProviderIdentifier
      *     The unique identifier of the AuthenticationProvider that created the
      *     UserContext being retrieved. Only one UserContext per
```
Comment (jmuehlner) [inline guacamole/src/main/java/org/glyptodon/guacamole/net/basic/rest/ObjectRetrievalService.java]: This is a little hard to understand. One UserContext per AuthenticationProvider? Surely that's per user?

File: guacamole/src/main/java/org/glyptodon/guacamole/net/basic/rest/ObjectRetrievalService.java [hunk 2]
```diff
@@ -110,6 +117,35 @@ public User retrieveUser(UserContext userContext,
...
 
     /**
+     * Retrieves a single user from the given GuacamoleSession.
+     *
+     * @param session
+     *     The GuacamoleSession to retrieve the user from.
+     *
+     * @param authProviderIdentifier
+     *     The unique identifier of the AuthenticationProvider that created the
+     *     UserContext from which the user should be retrieved. Only one
```
Comment (jmuehlner) [inline guacamole/src/main/java/org/glyptodon/guacamole/net/basic/rest/ObjectRetrievalService.java]: Same here

File: guacamole/src/main/java/org/glyptodon/guacamole/net/basic/rest/ObjectRetrievalService.java [hunk 3]
```diff
@@ -178,4 +244,37 @@ public ConnectionGroup retrieveConnectionGroup(UserContext userContext,
 
     }
 
+    /**
+     * Retrieves a single connection group from the given GuacamoleSession. If
+     * the given identifier the REST API root identifier, the root connection
```
Comment (jmuehlner) [inline guacamole/src/main/java/org/glyptodon/guacamole/net/basic/rest/ObjectRetrievalService.java]: I don't what "If  the given identifier the REST API root identifier" means.

File: guacamole/src/main/java/org/glyptodon/guacamole/net/basic/rest/ObjectRetrievalService.java [hunk 4]
```diff
@@ -178,4 +244,37 @@ public ConnectionGroup retrieveConnectionGroup(UserContext userContext,
...
+     *     The unique identifier of the AuthenticationProvider that created the
+     *     UserContext from which the connection group should be retrieved.
+     *     Only one UserContext per AuthenticationProvider can exist.
+     *
+     * @param identifier
+     *     The identifier of the connection group to retrieve.
+     *
+     * @return
+     *     The connection group having the given identifier, or the root
+     *     connection group if the identifier the root identifier.
```
Comment (jmuehlner) [inline guacamole/src/main/java/org/glyptodon/guacamole/net/basic/rest/ObjectRetrievalService.java]: You might want to this one also.

File: guacamole/src/main/java/org/glyptodon/guacamole/net/basic/rest/ObjectRetrievalService.java (hunk 3 above)
Comment (mike-jumper) [inline guacamole/src/main/java/org/glyptodon/guacamole/net/basic/rest/ObjectRetrievalService.java]: You're comment doesn't a verb.

File: guacamole/src/main/java/org/glyptodon/guacamole/net/basic/rest/activeconnection/ActiveConnectionRESTService.java [hunk 5]
```diff
@@ -140,6 +155,10 @@
      *     The authentication token that is used to authenticate the user
      *     performing the operation.
      *
+     * @param authProviderIdentifier
+     *     The unique identifier of the AuthenticationProvider associated with
+     *     the UserContext containing the active connections to be deleted.
```
Comment (jmuehlner) [inline guacamole/src/main/java/org/glyptodon/guacamole/net/basic/rest/activeconnection/ActiveConnectionRESTService.java]: What's being deleted here?

File: guacamole/src/main/java/org/glyptodon/guacamole/net/basic/rest/auth/APIAuthenticationResponse.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (C) 2014 Glyptodon LLC
```
Comment (jmuehlner) [inline guacamole/src/main/java/org/glyptodon/guacamole/net/basic/rest/auth/APIAuthenticationResponse.java]: (2015)

File: guacamole/src/main/webapp/app/manage/controllers/manageUserController.js [hunk 7]
```diff
@@ -123,54 +157,226 @@ angular.module('manage').controller('manageUserController', ['$scope', '$injecto
...
+        // Permissions can always be set if we are creating the user
+        if (!userExists)
+            return true;
+
+        // The administrator can always modify permissions
+        if (PermissionSet.hasSystemPermission($scope.permissions,
+                PermissionSet.SystemPermissionType.ADMINISTER))
+            return true;
 
+        // Otherwise, can only modify permissions if we have explicit ADMINSTER permission
```
Comment (jmuehlner) [inline guacamole/src/main/webapp/app/manage/controllers/manageUserController.js]: typo: "ADMINSTER"

File: guacamole/src/main/webapp/app/manage/controllers/manageUserController.js [hunk 8]
```diff
@@ -123,54 +157,226 @@ angular.module('manage').controller('manageUserController', ['$scope', '$injecto
...
+            return false;
+
+        // Only the administrator can modify system permissions
+        return PermissionSet.hasSystemPermission($scope.permissions,
+            PermissionSet.SystemPermissionType.ADMINISTER);
+
+    };
+
+    /**
+     * Returns whether the current user can save the user being edited,
```
Comment (jmuehlner) [inline guacamole/src/main/webapp/app/manage/controllers/manageUserController.js]: "user being edited, creating, or updating that user"?

File: guacamole/src/main/webapp/app/settings/directives/guacSettingsConnections.js [hunk 9]
```diff
@@ -111,20 +123,25 @@ angular.module('settings').directive('guacSettingsConnections', [function guacSe
...
-                    && $scope.permissions               !== null
-                    && $scope.canManageConnections      !== null
-                    && $scope.canCreateConnections      !== null
-                    && $scope.canCreateConnectionGroups !== null;
+                return $scope.rootGroup   !== null
+                    && $scope.permissions !== null;
 
             };
 
+            $scope.canManageConnections = true;
```
Comment (jmuehlner) [inline guacamole/src/main/webapp/app/settings/directives/guacSettingsConnections.js]: What's all this about?

File: guacamole/src/main/webapp/app/settings/directives/guacSettingsConnections.js [hunk 10]
```diff
@@ -111,20 +123,25 @@ angular.module('settings').directive('guacSettingsConnections', [function guacSe
...
+                permissionService.getPermissions,
+                [$scope.dataSource],
+                currentUsername
+            )
+            .then(function permissionsRetrieved(permissions) {
 
                 $scope.permissions = permissions;
-                                
+/*
                 // Ignore permission to update root group
```
Comment (jmuehlner) [inline guacamole/src/main/webapp/app/settings/directives/guacSettingsConnections.js]: Why is this commented out?

File: guacamole/src/main/webapp/app/settings/types/ActiveConnectionWrapper.js [hunk 11]
```diff
@@ -31,44 +31,47 @@ angular.module('settings').factory('ActiveConnectionWrapper', [
...
-     *     The ActiveConnection to wrap.
+     * @param {ActiveConnectionWrapper|Object} template
+     *     The object whose properties should be copied within the new
+     *     ActiveConnectionWrapper.
      */
-    var ActiveConnectionWrapper = function ActiveConnectionWrapper(name, startDate, activeConnection) {
+    var ActiveConnectionWrapper = function ActiveConnectionWrapper(template) {
+
+        /**
+         * The identifier of the data source associate dwith the
```
Comment (jmuehlner) [inline guacamole/src/main/webapp/app/settings/types/ActiveConnectionWrapper.js]: Typo here

File: guacamole/src/main/java/org/glyptodon/guacamole/net/basic/rest/ObjectRetrievalService.java (hunk 4 above)
Comment (mike-jumper) [inline guacamole/src/main/java/org/glyptodon/guacamole/net/basic/rest/ObjectRetrievalService.java]: Man ... why did I so many.

File: guacamole/src/main/webapp/app/settings/directives/guacSettingsConnections.js [hunk 12]
```diff
@@ -111,57 +123,142 @@ angular.module('settings').directive('guacSettingsConnections', [function guacSe
...
+
+                    // Can create connections groups if adminstrator or have explicit permission
+                    if (PermissionSet.hasSystemPermission(permissionSet, PermissionSet.SystemPermissionType.ADMINISTER)
+                     || PermissionSet.hasSystemPermission(permissionSet, PermissionSet.SystemPermissionType.CREATE_CONNECTION_GROUP))
+                        return dataSource;
+
+                }
+
+                // No data sources allow connection group creation
+                return null;
```
Comment (jmuehlner) [inline guacamole/src/main/webapp/app/settings/directives/guacSettingsConnections.js]: Did you mean false?

**Pull request merged**